### PR TITLE
Count pod overhead against an entity's ResourceQuota

### LINKED ISSUE #99577: PodOverhead resource not counted into ResourceQuota used info
#### What happened:

assume that: cluster with only one node with 4G allocatable memory

Define test runtimeclass with PodOverhead
```
overhead:
  podFixed:
    cpu: 50m
    memory: 200Mi
```
Define resourceQuota test under test namespace with 4Gi memory limit
```
status:
  hard:
    memory: "4Gi"
  used:
    memory: "0"
    pods: "0"
```
Create Pod with test runtimeClass under test namespace
```
    resources:
      limits:
        cpu: "1"
        memory: 2Gi
      requests:
        cpu: "1"
        memory: 2Gi
 ....
  overhead:
    cpu: 50m
    memory: 200Mi
```

podOverhead resource not counted into resourceQuota used info
```
kubectl get resourcequota test -n test -o yaml
status:
  hard:
    memory: "4Gi"
  used:
    memory: "2Gi"
    pods: "1"
```

podOverhead resource counted into node requests info
```
kubectl describe node ***

Namespace           Name                    CPU Requests  CPU Limits  Memory Requests  Memory Limits  AGE
 ---------              ----                        ------------  ----------  ---------------  -------------  ---
test                       testpod                   1050m (2%)    1050m (2%)  2248Mi (55%)      2248Mi (55%)    4d3h
```

try to create one more pod with 2G memory requests, resourceQuota check passed but failed to schedule because it requests 2G+200M memory

#### What you expected to happen:
podOverhead resource should be counted into resourceQuota used info to keep resourceQuota check and schedule check be consistent.

#### Environment:
- Kubernetes version 1.16

### PR DESCRIPTION
**Pull request by gjkim42**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

According to https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/688-pod-overhead, the ResourceQuota controller should account for `pod.Spec.Overhead`

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/99577

#### Does this PR introduce a user-facing change?

```release-note
Count pod overhead against an entity's ResourceQuota
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by gjkim42** [comment]
/cc @egernst 
Could you review this?

**Comment by gjkim42** [comment]
Need ok-to-test!

**Comment by chestack** [comment]
@gjkim42 thanks for your quick fix. 

About PodOverhead, I have something else to discuss. 

Currently, PodOverhead is defined as a fixed resource.
```
type Overhead struct {
  PodFixed *ResourceList
}
```
Actually, it's used as requests resource during scheduling. It looks like the average usage value. How about defining it like Pod resources including requests and limits. 
```
    resources:
      limits:
        memory: 1Gi
      requests:
        memory: 200M
```
Enable to configure limits(bigger than fixed resource value) is helpful to set the pod cgroup limit value when pod is running with high workload. 

My use case is kata_container, the VMM related processes consume not less overhead resource, a small overhead(small cgroup limits) will trigger oom but a big overhead result in that big overhead requests take into account during scheduling. 
So PodOverhead including limits and requests cloud be goog chioce.

**Comment by gjkim42** [comment]
> So how about define PodOverhead including limits and requests like that in Pod resources definition. It could be helpful
> 
> ```
>     resources:
>       limits:
>         memory: 1Gi
>       requests:
>         memory: 200M
> ```
@chestack 
In my opinion, there seem to be some places to be used. Maybe, you can make an issue to request api-change. However, I am not familiar with this domain. 

@egernst need your opinion!

**Comment by gjkim42** [comment]
Test failed because of https://github.com/kubernetes/kubernetes/issues/99642

**Comment by gjkim42** [comment]
@dims 
Awaiting your review.

File: pkg/quota/v1/evaluator/core/pods.go
Comment (ehashman) [inline pkg/quota/v1/evaluator/core/pods.go:355]: I see that there is a feature flag for this:

https://github.com/kubernetes/kubernetes/blob/267afe2358db44dd764b785744448e403359e5fe/pkg/features/kube_features.go#L738

Hence, we should probably wrap this logic with the feature flag.

**Review by ehashman — CHANGES_REQUESTED** [review]

File: pkg/quota/v1/evaluator/core/pods.go
Comment (gjkim42) [reply pkg/quota/v1/evaluator/core/pods.go]: Thanks. Done.

**Comment by dims** [comment]
/approve

leaving `/lgtm` for @ehashman

**Comment by gjkim42** [comment]
/hold
Let me do some tests.

**Review by ehashman — COMMENTED** [review]
Neat, thanks for the very quick turnaround!

/lgtm

**Comment by mrunalp** [comment]
Needs rebase

**Comment by gjkim42** [comment]
Rebased. 
/assign @derekwaynecarr @vishh (for approval)

**Comment by gjkim42** [comment]
Rebased. PTAL
/assign @vishh (for approval)
ping @derekwaynecarr

File: pkg/quota/v1/evaluator/core/pods_test.go
Comment (SergeyKanzhelev) [inline pkg/quota/v1/evaluator/core/pods_test.go:100]: so default is `false`? And all tests that do not set this field will execute with the pod overhead disabled? What am I missing?

**Comment by SergeyKanzhelev** [comment]
So this [KEP paragraph](https://github.com/kubernetes/enhancements/blob/7e69f3be9783e33f597b02de8c952b6eca1e8273/keps/sig-node/688-pod-overhead/README.md#resourcequota-changes) wasn't implemented?

KEP xref: #https://github.com/kubernetes/enhancements/issues/688

File: pkg/quota/v1/evaluator/core/pods_test.go
Comment (SergeyKanzhelev) [inline pkg/quota/v1/evaluator/core/pods_test.go:442]: can you please name the test case better? It doesn't test the quota changes, correct?

File: pkg/quota/v1/evaluator/core/pods_test.go
Comment (SergeyKanzhelev) [inline pkg/quota/v1/evaluator/core/pods_test.go:450]: no need for these. As asked above - do not mix test cases, it's better to test overhead explicitly

File: pkg/quota/v1/evaluator/core/pods_test.go
Comment (SergeyKanzhelev) [inline pkg/quota/v1/evaluator/core/pods_test.go:469]: this test case tests for usage, not for quota, correct? Sorry I might be missing something

Comment (gjkim42) [reply pkg/quota/v1/evaluator/core/pods_test.go:462]: According to [the KEP](https://github.com/kubernetes/enhancements/blob/7e69f3be9783e33f597b02de8c952b6eca1e8273/keps/sig-node/688-pod-overhead/README.md#resourcequota-changes) (`The controller will be updated to add the pod Overhead to the container resource request summation.`), I would say that `the container resource request summation` means the usage. And #99577  requested the same thing.

Yes, the test is for usage, and resourcequota controller ensures the usage to not exceed the quota

File: pkg/quota/v1/evaluator/core/pods_test.go
Comment (gjkim42) [inline pkg/quota/v1/evaluator/core/pods_test.go:499]: added the case containing pod overhead with the pod overhead disabled

**Review by gjkim42 — COMMENTED** [review]
@SergeyKanzhelev 
Done.

File: pkg/quota/v1/evaluator/core/pods_test.go
Comment (gjkim42) [reply pkg/quota/v1/evaluator/core/pods_test.go]: Done.

**Comment by gjkim42** [comment]
> So this [KEP paragraph](https://github.com/kubernetes/enhancements/blob/7e69f3be9783e33f597b02de8c952b6eca1e8273/keps/sig-node/688-pod-overhead/README.md#resourcequota-changes) wasn't implemented?

I don't know the full history, but I think it wasn't.

File: pkg/quota/v1/evaluator/core/pods_test.go
Comment (SergeyKanzhelev) [reply pkg/quota/v1/evaluator/core/pods_test.go:100]: got it. So there is no intention to explicitly disable it for other cases. It is just a side effect of the need to enable it for one case. Since it's in beta and about to get GA, does this logic even needed?

If you think it is needed, maybe just do nothing when it's false. See comment in file below

File: pkg/quota/v1/evaluator/core/pods_test.go
Comment (SergeyKanzhelev) [inline pkg/quota/v1/evaluator/core/pods_test.go:503]:
```suggestion
if (testCase.podOverheadEnabled) {
			defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.PodOverhead, testCase.podOverheadEnabled)()
			}
```

File: pkg/quota/v1/evaluator/core/pods_test.go
Comment (SergeyKanzhelev) [reply pkg/quota/v1/evaluator/core/pods_test.go:100]: oh, I see the new test case. it's ok than. Hopefully will GA PodOverhead in 1.22

**Comment by gjkim42** [comment]
@derekwaynecarr 
Need your approval

File: pkg/quota/v1/evaluator/core/pods.go
Comment (derekwaynecarr) [inline pkg/quota/v1/evaluator/core/pods.go:357]: i _wish_ we had this in beta as intended happy we caught it prior to moving to GA.

**Comment by derekwaynecarr** [comment]
/approve
/lgtm

this should probably warrant cherry-picks to past releases.

**Pull request merged**
